### PR TITLE
chore: internal request preflight

### PR DESCRIPTION
**Pull request by tgriesser**
### Additional details

Adds request preflight for Cypress Cloud recording.

Addresses https://github.com/cypress-io/engineering-documentation/pull/44

### Steps to test

Test that recording to Dashboard works.

### How has the user experience changed?

Doesn't directly impact user experience.

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/server/lib/cloud/api.ts [hunk 1]
```diff
@@ -11,7 +11,11 @@ const pkg = require('@packages/root')
 
 const machineId = require('./machine_id')
 const errors = require('../errors')
-const { apiRoutes } = require('./routes')
+const { apiUrl, apiRoutes, makeRoutes } = require('./routes')
+
+import Bluebird from 'bluebird'
```
Comment (ryanthemanuel) [inline packages/server/lib/cloud/api.ts:15]: `bluebird` is already required on line 5.

Comment (tgriesser) [reply packages/server/lib/cloud/api.ts:15]: Ah I see required as `Promise` - gonna update all `Promise` references in the file to be `Bluebird` so it's more explicit

File: packages/errors/src/errors.ts [hunk 2]
```diff
@@ -206,6 +206,12 @@ export const AllCypressErrors = {
         ${fmt.highlightSecondary(arg1.response)}`
   },
   CLOUD_UNKNOWN_CREATE_RUN_WARNING: (arg1: {props: any, message: string}) => {
```
Comment (emilyrohrbough) [inline packages/errors/src/errors.ts]:
```suggestion
  CLOUD_UNKNOWN_CREATE_RUN_WARNING: (arg1: {props?: any, message: string}) => {
```

Comment (tgriesser) [reply packages/errors/src/errors.ts]: Thanks!

File: packages/server/lib/cloud/api.ts [hunk 3]
```diff
@@ -353,5 +453,30 @@ module.exports = {
...
+        },
+        headers: {
+          'x-route-version': '1',
+          'x-cypress-request-attempt': attemptIndex,
+        },
+        json: true,
+        encrypt: 'always',
+      })
+
+      preflightResult = result // { encrypt: boolean, apiUrl: string }
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/api.ts:474]: Above we check for `always` string. Should this be boolean | 'always'

File: packages/server/lib/cloud/api.ts [hunk 4]
```diff
@@ -353,5 +453,30 @@ module.exports = {
     responseCache = {}
   },
 
+  preflight (preflightInfo) {
+    return retryWithBackoff(async (attemptIndex) => {
+      const preflightBase = process.env.CYPRESS_API_URL ? apiUrl.replace('api', 'api-proxy') : apiUrl
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/api.ts:458]: Why conditional based on process.env.CYPRESS_API_URL ?

File: packages/server/lib/cloud/api.ts [hunk 5]
```diff
@@ -30,20 +33,35 @@ const runnerCapabilities = {
...
-    return Promise.resolve(resp)
+    return Bluebird.resolve(resp)
   }
 
   _.defaults(params, {
     agent,
     proxy: null,
     gzip: true,
     cacheable: false,
+    encrypt: false,
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/api.ts:64]: Is there a link to the cases when this should be false?

File: packages/server/lib/cloud/api.ts [hunk 6]
```diff
@@ -144,9 +198,16 @@ const tagError = function (err) {
 }
 
 // retry on timeouts, 5xx errors, or any error without a status code
+// do not retry on decryption errors
 const isRetriableError = (err) => {
-  return (err instanceof Promise.TimeoutError) ||
-    (500 <= err.statusCode && err.statusCode < 600) ||
+  // .error is exposed as a property when the error is thrown
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/api.ts]:
```suggestion
  // error is exposed as a property when the error is thrown
```

File: packages/server/lib/cloud/encryption.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
+import crypto from 'crypto'
+import { TextEncoder, promisify } from 'util'
+import { generalDecrypt } from 'jose'
+import base64Url from 'base64url'
+import type { CypressRequestOptions } from './api'
+import { deflateRaw as deflateRawCb } from 'zlib'
+
+const deflateRaw = promisify(deflateRawCb)
+
+const keys: Record<string, string> = {
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/encryption.ts:10]: Should these keys be private keys pull from ENVs set in CI?

File: packages/server/lib/cloud/api.ts [hunk 8]
```diff
@@ -166,9 +227,32 @@ export type CreateRunOptions = {
   timeout?: number
 }
 
+let preflightResult = {
+  encrypt: true,
+  apiUrl,
+}
+
+let recordRoutes = apiRoutes
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/api.ts:235]: why set to local var? if easier to reason with, can we rename at the import level so this is consistent ?

File: system-tests/__snapshots__/record_spec.js [hunk 9]
```diff
@@ -2571,3 +2571,123 @@ Available browsers found on your system are:
...
+
+DecryptionError: JWE Recipients missing or incorrect type
+
+`
+
+exports['e2e record /preflight preflight failure: warning message renders preflight warning messages prior to run warnings 1'] = `
+Warning from Cypress Cloud: 
+
+----------------------------------------------------------------------
+This feature will not be supported soon, please check with Cypress to learn more: https://on.cypress.io/
```
Comment (emilyrohrbough) [inline system-tests/__snapshots__/record_spec.js:2613]: Should this wrap to fit in the `---- ` width?

File: packages/server/lib/cloud/api.ts [hunk 10]
```diff
@@ -187,39 +271,52 @@ module.exports = {
...
-          'ciBuildId',
-          'projectId',
-          'recordKey',
-          'specPattern',
-          'tags',
-          'testingType',
-        ]),
-        runnerCapabilities,
-      }
+    const preflightOptions = _.pick(options, ['projectId', 'ciBuildId', 'browser'])
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/api.ts:274]: Should this be pulling the Cypress version to ensure record compatibility?

File: packages/server/lib/cloud/api.ts [hunk 11]
```diff
@@ -353,5 +453,30 @@ module.exports = {
...
+        headers: {
+          'x-route-version': '1',
+          'x-cypress-request-attempt': attemptIndex,
+        },
+        json: true,
+        encrypt: 'always',
+      })
+
+      preflightResult = result // { encrypt: boolean, apiUrl: string }
+      recordRoutes = makeRoutes(result.apiUrl)
```
Comment (emilyrohrbough) [inline packages/server/lib/cloud/api.ts:475]: Why would the route be changing vs using our record routes with some sort of auth token on requests?

File: scripts/after-pack-hook.js [hunk 12]
```diff
@@ -58,8 +58,17 @@ module.exports = async function (params) {
...
 
-      await fs.writeFile(path.join(outputFolder, 'index.js'), binaryEntryPointSource)
+      if (!fileContents.includes(`test: CY_TEST,`)) {
+        throw new Error(`Expected to find test key in cloud encryption file`)
+      }
+
+      await Promise.all([
+        fs.writeFile(encryptionFile, fileContents.replace(`test: CY_TEST,`, '')),
+        fs.writeFile(path.join(outputFolder, 'index.js'), binaryEntryPointSource),
+      ])
```
Comment (brian-mann) [inline scripts/after-pack-hook.js]: we should add another assertion here that `fileContents` does not contain `CY_TEST` after writing it to verify the stripping

File: scripts/after-pack-hook.js [hunk 13]
```diff
@@ -70,6 +79,8 @@ module.exports = async function (params) {
         },
       )
 
+      process.env.V8_UPDATE_METAFILE = '1'
```
Comment (ryanthemanuel) [inline scripts/after-pack-hook.js:88]: Thanks for fixing this

File: .github/workflows/update_v8_snapshot_cache.yml [hunk 14]
```diff
@@ -5,6 +5,7 @@ on:
     - cron: '0 0 * * 3'
   push:
     branches:
+      # - tgriesser/spike/spike
```
Comment (mjhenkes) [inline .github/workflows/update_v8_snapshot_cache.yml]: remove?

File: .circleci/workflows.yml [hunk 15]
```diff
@@ -28,7 +28,7 @@ mainBuildFilters: &mainBuildFilters
       only:
         - develop
         - /^release\/\d+\.\d+\.\d+$/
-        - 'ryanm/fix/issues-with-v8-snapshot-on-develop'
+        - 'tgriesser/spike/_spike'
```
Comment (ryanthemanuel) [inline .circleci/workflows.yml]: Is there a reason we have the `_` in _spike here? We're not actually testing the other platforms here.

Comment (ryanthemanuel) [inline .circleci/workflows.yml]:
```suggestion
        - 'tgriesser/spike/spike'
```

File: .circleci/workflows.yml [hunk 16]
```diff
@@ -37,7 +37,7 @@ macWorkflowFilters: &darwin-workflow-filters
   when:
     or:
     - equal: [ develop, << pipeline.git.branch >> ]
-    - equal: [ 'ryanm/fix/issues-with-v8-snapshot-on-develop', << pipeline.git.branch >> ]
+    - equal: [ 'tgriesser/spike/_spike', << pipeline.git.branch >> ]
```
Comment (ryanthemanuel) [inline .circleci/workflows.yml]:
```suggestion
    - equal: [ 'tgriesser/spike/spike', << pipeline.git.branch >> ]
```

File: .circleci/workflows.yml [hunk 17]
```diff
@@ -130,7 +130,7 @@ commands:
       - run:
           name: Check current branch to persist artifacts
           command: |
-            if [[ "$CIRCLE_BRANCH" != "develop" && "$CIRCLE_BRANCH" != "release/"* && "$CIRCLE_BRANCH" != "ryanm/fix/issues-with-v8-snapshot-on-develop" ]]; then
+            if [[ "$CIRCLE_BRANCH" != "develop" && "$CIRCLE_BRANCH" != "release/"* && "$CIRCLE_BRANCH" != "tgriesser/spike/_spike" ]]; then
```
Comment (ryanthemanuel) [inline .circleci/workflows.yml]:
```suggestion
            if [[ "$CIRCLE_BRANCH" != "develop" && "$CIRCLE_BRANCH" != "release/"* && "$CIRCLE_BRANCH" != "tgriesser/spike/spike" ]]; then
```

File: .circleci/workflows.yml [hunk 18]
```diff
@@ -46,7 +46,7 @@ linuxArm64WorkflowFilters: &linux-arm64-workflow-filters
   when:
     or:
     - equal: [ develop, << pipeline.git.branch >> ]
-    - equal: [ 'ryanm/fix/issues-with-v8-snapshot-on-develop', << pipeline.git.branch >> ]
+    - equal: [ 'tgriesser/spike/_spike', << pipeline.git.branch >> ]
```
Comment (ryanthemanuel) [inline .circleci/workflows.yml]:
```suggestion
    - equal: [ 'tgriesser/spike/spike', << pipeline.git.branch >> ]
```

File: .github/workflows/update_v8_snapshot_cache.yml (hunk 14 above)
Comment (mjhenkes) [reply .github/workflows/update_v8_snapshot_cache.yml]:
```suggestion
```

**Review by ryanthemanuel — APPROVED** [review]

**Review by mjhenkes — APPROVED** [review]
Looks good, I wouldn't mind seeing more code doc and comments in the encryption.ts file.

File: packages/server/package.json [hunk 19]
```diff
@@ -36,6 +36,7 @@
     "@packages/icons": "0.0.0-development",
     "ansi_up": "5.0.0",
     "ast-types": "0.13.3",
+    "base64url": "^3.0.1",
```
Comment (flotwig) [inline packages/server/package.json:39]: Is it necessary to add a new dependency? AFAICT everything's running on the server, where we can use `Buffer.from(str, 'base64')` and `Buffer.toString('base64')` to decode/encode base64.

**Review by flotwig — APPROVED** [review]

File: packages/server/package.json (hunk 19 above)
Comment (flotwig) [reply packages/server/package.json:39]: I guess here it's used to convert between two base64 encoding schemes?

https://github.com/cypress-io/cypress/blob/6ca21d6a8755895302eb4fc285ced1664e720ed7/packages/server/test/unit/cloud/api_spec.js#L61

Comment (tgriesser) [reply packages/server/package.json:39]: Yeah, `base64url` is slightly different than `base64`. I think support for it landed in 14.16 but ran into issues bumping `@types/node` to not complain about it. The code for it [is pretty simple](https://github.com/brianloveswords/base64url/blob/fce104a75ebade9bbeb761d74153b46938c2fef4/src/base64url.ts#L14-L28), I had orignially inlined but then didn't think it was worth porting the unit tests for it vs just having it as a dependency. We can remove it later when we bump the `@types/node` and it's a proper encoding target

File: .circleci/workflows.yml (hunk 15 above)
Comment (tgriesser) [reply .circleci/workflows.yml]: Yeah I had added it temporarily and then later added the `_` so it wasn't running everything on every commit once I verified what I needed

File: scripts/after-pack-hook.js (hunk 12 above)
Comment (tgriesser) [reply scripts/after-pack-hook.js]: Added

**Pull request merged**